### PR TITLE
Feature/add contact with due date

### PR DESCRIPTION
**Pull request by Amet015**

### CODE REVIEW

File: src/main/java/salesforce/entities/Contact.java [hunk 1]
```diff
@@ -13,6 +13,7 @@
 import com.fasterxml.jackson.annotation.JsonInclude;
 import salesforce.entities.constants.ContactConstant;
 import salesforce.utils.DateFormatter;
+
```
Comment (JuanCarlosMartinez4) [inline src/main/java/salesforce/entities/Contact.java]: delete this line

File: src/main/java/salesforce/entities/Opportunity.java [hunk 2]
```diff
@@ -13,6 +13,7 @@
 import com.fasterxml.jackson.annotation.JsonInclude;
 import salesforce.entities.constants.OpportunityConstant;
 import salesforce.utils.DateFormatter;
+
```
Comment (JuanCarlosMartinez4) [inline src/main/java/salesforce/entities/Opportunity.java]: delete and in all classes

File: src/main/java/salesforce/entities/TaskOpportunity.java [hunk 3]
```diff
@@ -72,17 +74,17 @@ public void setPriority(final String priority) {
      *
      * @return string value.
```
Comment (JuanCarlosMartinez4) [inline src/main/java/salesforce/entities/TaskOpportunity.java]: string value.? or contact value.

File: src/main/java/salesforce/ui/pages/opportunity/taskopportunity/AbstractTaskOpportunity.java [hunk 4]
```diff
@@ -25,14 +25,25 @@
...
+    /**
+     * Sets contact.
+     * @param contact value.
+     */
+    protected abstract void setContact(String contact);
+
+    /**
+     * Sets due date.
+     * @param dueDate value.
+     */
```
Comment (JuanCarlosMartinez4) [inline src/main/java/salesforce/ui/pages/opportunity/taskopportunity/AbstractTaskOpportunity.java]: ctrl alt l

File: src/main/java/salesforce/ui/pages/opportunity/taskopportunity/TaskOpportunityClassic.java [hunk 5]
```diff
@@ -41,6 +41,16 @@ protected void setSubjectTask(final String subject) {
...
 
+    @Override
+    protected void setContact(final String contact) {
+
+    }
+
+    @Override
+    protected void setDueDate(final String dueDate) {
+
+    }
```
Comment (JuanCarlosMartinez4) [inline src/main/java/salesforce/ui/pages/opportunity/taskopportunity/TaskOpportunityClassic.java:52]: if you do not implement put //To Do

File: src/main/java/salesforce/ui/pages/opportunity/taskopportunity/TaskOpportunityLightning.java [hunk 6]
```diff
@@ -32,16 +33,57 @@
...
+     * Selects a contact on the combo box.
+     * @param contact value.
+     */
+    private void selectContact(final String contact) {
+        String selectContactXpath = String.format(CONTACT_SPECIFIC, contact);
+        contactSelected = webDriver.findElement(By.xpath(selectContactXpath));
+        contactSelected.click();
+    }
+
+    /**
```
Comment (JuanCarlosMartinez4) [inline src/main/java/salesforce/ui/pages/opportunity/taskopportunity/TaskOpportunityLightning.java]: ctrl alt l

File: src/main/java/salesforce/ui/pages/task/AbstractTask.java [hunk 7]
```diff
@@ -84,6 +84,34 @@
      */
     protected abstract String getSubject(String subject);
 
+    /**
+     * Sets Contact.
+     *
+     * @param contact value.
+     */
+    protected abstract void setContact(String contact);
```
Comment (JuanCarlosMartinez4) [inline src/main/java/salesforce/ui/pages/task/AbstractTask.java:92]: **final** String contact

File: src/main/java/salesforce/ui/pages/task/AbstractTask.java [hunk 8]
```diff
@@ -84,6 +84,34 @@
...
+     *
+     * @return string value.
+     */
+    protected abstract String getContact();
+
+    /**
+     * Sets due date.
+     * @param dueDate value.
+     */
+    protected abstract void setDueDate(String dueDate);
```
Comment (JuanCarlosMartinez4) [inline src/main/java/salesforce/ui/pages/task/AbstractTask.java:105]: the same

File: src/main/java/salesforce/ui/pages/task/TaskClassicPage.java [hunk 9]
```diff
@@ -98,4 +98,24 @@ protected String getSubject(final String subject) {
         subjectName = webDriver.findElement(By.xpath(subjectXpath));
         return subjectName.getText();
     }
+
+    @Override
+    protected void setContact(final String contact) {
+
```
Comment (JuanCarlosMartinez4) [inline src/main/java/salesforce/ui/pages/task/TaskClassicPage.java:104]: put //To Do

File: src/main/java/salesforce/utils/ReplacerMessages.java [hunk 10]
```diff
@@ -32,4 +33,15 @@ public static String replaceChangeOwnerMessage(final String message, final Strin
         result = message.replace(CURRENT_USER_REGEX, jsonFileReader.getUser(CURRENT_USER).getUsername());
         return result.replace(OPPORTUNITY_NAME_REGEX, opportunityName);
     }
+
+
```
Comment (JuanCarlosMartinez4) [inline src/main/java/salesforce/utils/ReplacerMessages.java]: delete this line

File: src/main/java/salesforce/utils/ReplacerMessages.java [hunk 11]
```diff
@@ -32,4 +33,15 @@ public static String replaceChangeOwnerMessage(final String message, final Strin
...
+    /**
+     * Replaces words in a message.
+     * @param message contains a message String.
+     * @param taskName contains an opportunity name.
+     * @return a new message.
+     */
+    public static String replaceTaskSavedMessage(final String message, final String taskName) {
+        return message.replace(TASK_NAME, taskName);
```
Comment (JuanCarlosMartinez4) [inline src/main/java/salesforce/utils/ReplacerMessages.java]: ctrl alt l

File: src/test/java/salesforcetest/hooks/Hook.java [hunk 12]
```diff
@@ -73,18 +73,25 @@ public void deletesOpportunity() {
         }
     }
 
-    /**
+
```
Comment (JuanCarlosMartinez4) [inline src/test/java/salesforcetest/hooks/Hook.java]: delete this line

File: src/test/java/salesforcetest/hooks/Hook.java [hunk 13]
```diff
@@ -73,18 +73,25 @@ public void deletesOpportunity() {
...
     public void deletesLeads() {
         leadHelper.deleteLeads(context.getLeads());
     }
 
-     /**
-     * Deletes contacts by id.
+//     /**
+//     * Deletes contacts by id.
+//     */
+//    @After("@CampaignAccounts")
```
Comment (JuanCarlosMartinez4) [inline src/test/java/salesforcetest/hooks/Hook.java]: if you are not using, delete it

File: src/test/java/salesforcetest/steps/OpportunityStep.java [hunk 14]
```diff
@@ -236,18 +238,23 @@ public void selectOpportunity() {
...
     @And("I add new Task with")
-    public void iAddNewTaskWith(final Map<String, String> mapNewTask) {
+    public void iAddNewTaskWith(final Map<String, String> mapTask) {
+        mapNewTask = new HashMap<>(mapTask);
         opportunitiesPage = AppPageFactory.getTabObjectsPage();
         opportunityPage = opportunitiesPage.selectObjectByName(context.getOpportunities().get(0).getName());
         abstractTaskOpportunity = opportunityPage.clickAddTask();
         taskOpportunity = context.getTaskOpportunity();
+        if (!context.getContacts().isEmpty()) {
+            String contact = context.getContacts().get(0).getFirstName() + " " + context.getContacts().get(0)
```
Comment (JuanCarlosMartinez4) [inline src/test/java/salesforcetest/steps/OpportunityStep.java]: try ti avoid use magic numbers

File: src/test/java/salesforcetest/steps/OpportunityStep.java [hunk 15]
```diff
@@ -268,7 +277,23 @@ public void iAddAdditionalInformationToTheTask(final Map<String, String> mapAddI
...
     public void theTaskShouldDisplayTheInformationAdded() {
-        HashMap<String, String> mapTaskValidate = abstractTask.getTaskDetails(taskOpportunity);
+        HashMap<String, String> mapTaskValidate = abstractTask.getTaskDetails(taskOpportunity, mapNewTask.keySet());
         Assert.assertEquals(mapTaskValidate, context.getTaskOpportunity().getTaskEdited());
     }
+
+    /**
+     * Validates a message only for Lightning User Experience.
+     * @param message list.
+     */
```
Comment (JuanCarlosMartinez4) [inline src/test/java/salesforcetest/steps/OpportunityStep.java]: ctrl alt l

File: src/test/resources/features/Opportunity.feature [hunk 16]
```diff
@@ -4,39 +4,53 @@ Feature: Opportunity
   Scenario: Change opportunity owner to another user
     Given I create 1 "Basic" opportunities
     When I go to "Opportunities Page"
-      And I search for the opportunity in list "All Opportunities"
-      And I change the opportunity's owner with "Current User"
+    And I search for the opportunity in list "All Opportunities"
+    And I change the opportunity's owner with "Current User"
```
Comment (JuanCarlosMartinez4) [inline src/test/resources/features/Opportunity.feature]: tab

File: src/test/resources/features/Opportunity.feature [hunk 17]
```diff
@@ -4,39 +4,53 @@ Feature: Opportunity
...
     Then the application should display this message in Opportunity Page only for Lightning Experience
       | [Current User] now owns the record for [Opportunity Name]. |
-      And the opportunity page displays the owner "Current User"
+    And the opportunity page displays the owner "Current User"
```
Comment (JuanCarlosMartinez4) [inline src/test/resources/features/Opportunity.feature]: tab

File: src/main/java/salesforce/ui/pages/opportunity/taskopportunity/TaskOpportunityClassic.java (hunk 5 above)
Comment (Amet015) [reply src/main/java/salesforce/ui/pages/opportunity/taskopportunity/TaskOpportunityClassic.java:52]: why! now I am gonna work in it

File: src/main/java/salesforce/ui/pages/task/AbstractTask.java (hunk 7 above)
Comment (Amet015) [reply src/main/java/salesforce/ui/pages/task/AbstractTask.java:92]: this is abstract

File: src/main/java/salesforce/ui/pages/task/AbstractTask.java (hunk 8 above)
Comment (Amet015) [reply src/main/java/salesforce/ui/pages/task/AbstractTask.java:105]: me too, the same

File: src/main/java/salesforce/ui/pages/task/TaskClassicPage.java (hunk 9 above)
Comment (Amet015) [reply src/main/java/salesforce/ui/pages/task/TaskClassicPage.java:104]: now I'm gonna work in it

File: src/main/java/salesforce/ui/pages/task/AbstractTask.java (hunk 7 above)
Comment (JuanCarlosMartinez4) [reply src/main/java/salesforce/ui/pages/task/AbstractTask.java:92]: check again

File: src/main/java/salesforce/ui/pages/task/AbstractTask.java (hunk 8 above)
Comment (JuanCarlosMartinez4) [reply src/main/java/salesforce/ui/pages/task/AbstractTask.java:105]: x2

File: src/main/java/salesforce/utils/DateFormatter.java [hunk 18]
```diff
@@ -157,4 +161,22 @@ public static String formatDate(final String date) {
         }
         return value;
     }
+
+    /**
+     * Converts format date to format to use on UI.
```
Comment (JuanCarlosMartinez4) [inline src/main/java/salesforce/utils/DateFormatter.java:166]: check this please

File: src/main/java/salesforce/utils/ReplacerMessages.java [hunk 19]
```diff
@@ -19,10 +19,12 @@
     private static final String CURRENT_USER_REGEX = "[Current User]";
     private static final String OPPORTUNITY_NAME_REGEX = "[Opportunity Name]";
     private static final String CURRENT_USER = "Current User";
+    private static final String TASK_NAME = "[Task Name]";
 
     /**
      * Replaces words in a message.
-     * @param message contains a message String.
+     *
+     * @param message         contains a message String.
```
Comment (JuanCarlosMartinez4) [inline src/main/java/salesforce/utils/ReplacerMessages.java]: delete **param message         contains a me** the space

File: src/test/resources/features/Opportunity.feature (hunk 17 above)
Comment (JuanCarlosMartinez4) [reply src/test/resources/features/Opportunity.feature]: x2

**Review by JuanCarlosMartinez4 — COMMENTED** [review]
check all observations please

File: src/main/java/salesforce/ui/pages/task/AbstractTask.java (hunk 7 above)
Comment (Amet015) [reply src/main/java/salesforce/ui/pages/task/AbstractTask.java:92]: this is class abstract!!! checkstyle doesn't display error!

File: src/main/java/salesforce/ui/pages/task/AbstractTask.java (hunk 8 above)
Comment (Amet015) [reply src/main/java/salesforce/ui/pages/task/AbstractTask.java:105]: this is class abstract!!! checkstyle doesn't display error!

File: src/main/java/salesforce/utils/DateFormatter.java (hunk 18 above)
Comment (Amet015) [reply src/main/java/salesforce/utils/DateFormatter.java:166]: what is you propouse?

File: src/test/java/salesforcetest/steps/OpportunityStep.java [hunk 20]
```diff
@@ -236,18 +239,25 @@ public void selectOpportunity() {
...
         opportunitiesPage = AppPageFactory.getTabObjectsPage();
-        opportunityPage = opportunitiesPage.selectObjectByName(context.getOpportunities().get(0).getName());
+        opportunityPage = opportunitiesPage.selectObjectByName(context.getOpportunities().get(ARRAY_POSITION_FIRST)
+                .getName());
         abstractTaskOpportunity = opportunityPage.clickAddTask();
         taskOpportunity = context.getTaskOpportunity();
+        if (!context.getContacts().isEmpty()) {
+            String contact = context.getContacts().get(ARRAY_POSITION_FIRST).getFirstName() + " " + context
+                    .getContacts().get(ARRAY_POSITION_FIRST)
+                    .getLastName();
```
Comment (JuanCarlosMartinez4) [inline src/test/java/salesforcetest/steps/OpportunityStep.java]: this is no necessary on new line

**Review by JuanCarlosMartinez4 — APPROVED** [review]

**Pull request merged**